Themis/ViewModels/Assessment/AssessmentViewModel.swift

```diff
@@ -8,37 +8,132 @@ class AssessmentViewModel: ObservableObject {
+    func resetForNewAssessment() {
+        self.submission = nil
+        self.participation = nil
+        self.loading = false
+        self.error = nil
+        self.submissionId = nil
+        self.participationId = nil
+        self.resultId = nil
+    }
+    
+    /// Resets some properties of this viewmodel that affect the toolbar
+    func resetToolbarProperties() {
+        pencilMode = true
+        fontSize = 16.0
     }
     
     var gradingCriteria: [GradingCriterion] {
         participation?.getExercise()?.gradingCriteria ?? []
     }
+    
+    @MainActor
+    func initSubmission() async {
+        guard submission == nil else {
+            return
+        }
+        
+        switch exercise {
+        case .programming(exercise: _):
+            if readOnly {
+                if let participationId {
+                    await getReadOnlySubmission(participationId: participationId)
+                }
+            } else {
+                if let submissionId {
+                    await getSubmission(submissionId: submissionId)
+                } else {
+                    await initRandomSubmission()
+                }
+            }
+        case .text(exercise: _):
+            if readOnly {
+                // TODO: figure out which endpoint could be used instead
+            } else {
+                if let participationId, let submissionId {
+                    await getParticipationForSubmission(participationId: participationId, submissionId: submissionId)
+                } else {
+                    await initRandomSubmission()
+                }
+            }
+        default:
+            log.warning("Attempt to assess an unknown exercise")
+        }
+        
+        ThemisUndoManager.shared.removeAllActions()
+    }
+    
+    @MainActor
+    private func getParticipationForSubmission(participationId: Int?, submissionId: Int?) async {
+        guard let participationId, let submissionId else {
+            return
+        }
+        
+        loading = true
+        defer {
+            loading = false
+        }
+        do {
+            let assessmentService = AssessmentServiceFactory.service(for: exercise)
+            let fetchedParticipation = try await assessmentService.fetchParticipationForSubmission(participationId: participationId,
+                                                                                                   submissionId: submissionId).baseParticipation
+            self.submission = fetchedParticipation.submissions?.last?.baseSubmission
+            self.participation = fetchedParticipation
+            assessmentResult.setComputedFeedbacks(basedOn: participation?.results?.last?.feedbacks ?? [])
+            assessmentResult.setReferenceData(basedOn: submission)
+            ThemisUndoManager.shared.removeAllActions()
+        } catch {
+            self.submission = nil
+            self.error = error
+            log.info(String(describing: error))
```

Comment:
Why not error log here?

```suggestion
            log.error(String(describing: error))
```